### DISCUSSION TITLE
Support the minimum setpoint deadband on the climate entity.

### DISCUSSION THREAD

**Discussion by dbolger** [home-assistant/architecture#996] [Entity Models] [answered]
I've proposed in the frontend that we introduce some changes so the user knows what they can set their thermostat's temperature to. 

For example, on my Nest Thermostat, I have a 3°F minimum setpoint deadband (MSD). Setting my thermostat on heat/cool mode with the heat target at 70°F and the cool target at 73°F is valid. Anything that reduces the range between those two numbers is invalid (in the heat/cool mode). I drew some rough mock-ups of what supporting the MSD looks like, but it will require a variable that *should* exist on most thermostats with a heat/cool setting. Those markers on the climate slider could be beneficial to preventing users from inputting invalid values (and subsequently, in my case, throwing exceptions)

https://github.com/home-assistant/frontend/discussions/18456

**Comment by dbolger** [comment]
If we do go this route, perhaps it would be better to name the attribute something a bit different in the code, such as `MIN_HEAT_COOL_RANGE`, since MSD isn't particularly described well around the internet. Or we could add a code comment describing what MSD means.

**Comment by allenporter** [comment]
My impression is that right now what happens is that if you set the low end within that range it bumps the high range to "make room" so to speak after https://github.com/home-assistant/core/pull/98238 -- that is, there is no dead band, it just shifts the other side of the range to make it fit.

**Reply by dbolger** [reply]
Ah okay. Maybe the Matter integration's climate entity is neglecting to set/use that value. I'll review your changes you linked further and try to find out.

**Reply by allenporter** [reply]
There is a little more on the history here:  Some discussion in #69614 about how best to handle this went back and forth for a bit and the suggestion by the user in #97211 was to just handle it this way, which I thought was a pretty elegant idea.

Makes sense to see if the same idea could be used in matter too. I liked their solution since it didn't require any additional changes to our APIs or entity model.

**Reply by dbolger** [reply]
Wouldn't it be better to make this attribute a part of the climate entity considering more of them are starting to support it?

**Reply by marcelveldt** [reply]
I agree if this is something that is getting more common, we should solve it in a generic way. If its just a few integrations, we could just solve it at integration level. Imo the big benefit of having the minimum deadband as official climate property, it would mean the frontend can also limit the value to the user instead of an error when you try to set something outside the deadband.

**Reply by dbolger** [reply]
@frenck moving the conversation from the PR here:
What do you think about adding this attribute? This discussion has been open for two weeks.

**Reply by frenck** [reply]
This proposal misses one important factor: The usage.

Which integration can provide this property, as in, today?

As Home Assistant provides an abstraction, we try to facilitate the most common cases, and not the exceptions to this case. So, please provide a list of integrations that can leverage this new feature you are proposing.

../Frenck

**Reply by dbolger** [reply]
> This proposal misses one important factor: The usage.
> 
> Which integration can provide this property, as in, today?
> 
> As Home Assistant provides an abstraction, we try to facilitate the most common cases, and not the exceptions to this case. So, please provide a list of integrations that can leverage this new feature you are proposing.
> 
> ../Frenck

@frenck The usage here would be in the Nest integration's thermostat component and the Matter integration's thermostat component.

Currently, the Nest thermostat integration supports the deadband but the Matter integration does not. The idea here was to add the deadband to the climate entity model and then refactor the Nest integration's thermostat component to use the new attribute, and then implement the attribute in the Matter implementation. I can't name any more devices/integrations that have a heat/cool mode, however, the minimum setpoint deadband is a relatively common feature across devices. Searching Google for 'minimum setpoint deadband' shows it's a common thing.

A quick Google search shows it may be implemented separately in Zigbee as well:
https://community.home-assistant.io/t/zigbee-thermostat-centralite-pearl-thermostat/325795/80

**Reply by dbolger** [reply]
> This proposal misses one important factor: The usage.
> 
> 
> 
> Which integration can provide this property, as in, today?
> 
> 
> 
> As Home Assistant provides an abstraction, we try to facilitate the most common cases, and not the exceptions to this case. So, please provide a list of integrations that can leverage this new feature you are proposing.
> 
> 
> 
> ../Frenck

@frenck would you suggest to do it on the individual integration level until the need is higher?

**Reply by bendikrb** [reply]
My zwave based thermostats (under-floor heating) also has low and high temperature settings, in addition to set temp.

**Reply by dbolger** [reply]
@frenck @marcelveldt What's the latest on this?

**Comment by frenck** [comment]
We discussed this in our core meeting and agreed to accept this architectural proposal.

Even though the integration support at this point, would be very small, it does involve integration with a generic protocol like Z-Wave and Matter.

Also, the deadband could be enforced/included by the integration itself already; using this proposal could improve the UI experience and automatically do the enforcement for supporting integrations in the core entity component implementation (instead of them implementing it the integration logic).

What is missing in this proposal is how this technical implementation would look.

For example, this will need additional properties; which would those be? Is that something an integration need to specifically opt into? (does it need a feature flag?) Should it be added to our entity description? How would the UI use this and behave? Would it push and remain the gap? Or block? 

../Frenck

**Reply by dbolger** [reply]
In terms of additional properties, I think we'll need to define an attribute in the climate entity for the minimum gap in HEAT and COOL values. We could name the attribute `MIN_HEAT_COOL_RANGE`. I believe this should be added to the entity description as an attribute due to it's read-only nature.

I think for the best user experience, the gap should remain and depending on temperature setting changed, the opposing value should be pushed to keep the gap consistent. For example, say the minimum range is 3°F. Before making changes, the cool temperature is set to 68°F and the heat temperature is set to 71°F. If the user raises the cool temperature by 1°F, then I'd expect the heat temperature to automatically change to 72°F to accommodate for the gap. This should also happen if the user lowers the heat temperature by 1°F - we would see the cool temperature go to 67°F on the UI to maintain the gap. 

I think we may benefit the most if we add some sort of read-only conditional (in terms of user control) that controls whether the UI should automatically adjust for the gap or not. The reason I think we should add this is to ensure that only climate entities that have explicitly provided a minimum heat/cool range will actually handle the gap. This will ensure that we aren't forcing a gap on users if there wasn't one explicitly loaded in via the integration they are using. For example, my Nest Thermostat on the Matter integration **does** support getting the minimum range, so in that case we'd expect the UI to automatically handle the gap. However, if the integration does not support loading the minimum range (either because it has yet to be implemented, or because the device itself does not provide an interface in which the integration can use), then I would expect the UI to allow any values to be provided, without trying to automatically accommodate for a default gap, since defining a default gap could hinder the user from using their device as expected. An example of that would be us defining a minimum gap of 3°F but the hardware actually supporting 2°F, but not reporting it into Home Assistant. We'd be limiting the user more than they need to be. I think it should be the integration's responsibility to use the given attribute effectively. Integrations can slowly add support for the new attribute as more hardware begins to support it.

**Reply by marcelveldt** [reply, marked as answer]
Just for the record: I'm not really up to speed in thermostats and deadbands but the proposal looks fine to me and I like to have it implemented in Matter. So from a Matter perspective this proposal is 100% approved. Now, let's go ahead and implement it :-)

**Reply by emontnemery** [reply]
That makes sense @marcelveldt, this proposal is approved, let's indeed go ahead with the implementation 👍

**Reply by dbolger** [reply]
I think it's worth clarifying here - I don't think that if the user is trying to 'extend' the current range the minimum should be maintained. It should only be maintained if the gap is getting smaller and it reaches the minimum range value.

i.e The user has a low temperature set to 70F and a high set to 73F. If the user sets the low to 69F, the high temperature should not lower to 72F. High should stay at 73F since the gap is growing, not getting smaller. Say the low temperature changes to be set to 71F in the example instead of 69F. In this case, the UI should read the attribute and determine the new high temperature (in the case of my thermostat, 3F is the gap, so the new high would be 74F).

**Reply by emontnemery** [reply]
@dbolger can your concerns be worked out in an implementation PR, or does it need to be clarified in the architecture proposal?

**Reply by dbolger** [reply]
@emontnemery I believe those concerns should be handled by the frontend. I've talked to @marcelveldt and ideally the frontend should read the attribute if it's present to allow/disallow a change to either the heat or cool temperature. I don't know if that warrants another architecture proposal. I plan on implementing a check in the Matter implementation as a safeguard that if someone tries to send a climate set temperature request where the distance between the low and high temperatures is smaller than the minimum gap, an exception is thrown. However that exception shouldn't control the UI, it's only a means for preventing someone from trying to provide an incorrect pair of values.

**Reply by dbolger** [reply]
I updated my comment from yesterday to try and further clarify this.

**Reply by gjohansson-ST** [reply]
I think this would be easiest to just handle in the `async_service_temperature_set()` method within `climate` to ensure the deadband before it's passing the call onto the respective entity.
For visualization (climate card) if we add this an attribute it would be easy to read for frontend so the card can be updated (but it's not necessary to achieve the functionality as such).

**Reply by gjohansson-ST** [reply]
@dbolger Do you want to make a PR to add this functionality?

**Reply by dbolger** [reply]
@gjohansson-ST I will!

**Comment by dbolger** [comment]
For reference on the UI design: this is how the Google Home app responds to temperature changes.

https://github.com/home-assistant/architecture/assets/37427166/100a7435-d67c-4c87-af8b-c3b327ef5979
